llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp

```diff
@@ -4706,6 +4706,15 @@ Instruction *InstCombinerImpl::visitXor(BinaryOperator &I) {
     return Xor;
 
   Value *X, *Y;
+
+  // (A | B) ^ C -> (A ^ C) ^ B
+  // C ^ (A | B) -> B ^ (A ^ C)
+  if (match(&I, m_c_Xor(m_OneUse(m_c_DisjointOr(m_Value(X), m_Value(Y))),
+                        m_Value(M)))) {
+    Value *XorAC = Builder.CreateXor(X, M);
+    return BinaryOperator::CreateXor(XorAC, Y);
+  }
```

Comment:
Is this desirable? I'd say `disjoint or` is preferable to an `xor` in most if not all respects.

I would limit this if `X ^ M` folds. You can test it with:
`if (Value * XorAC = simplifyBinOp(Instruction::Xor, X, M, SQ)`

Reply:
You could also try if `Y, M` simplify.

Reply:
So can we do

if X ^ M folds or if Y ^ M folds -> (X ^ M) ^ Y

Reply:
Now i checked simplify and update samples, this is alive2 for current test depending that A ^ M will folded

https://alive2.llvm.org/ce/z/NLhfhN

Reply:
With `simplifyBinOp`, it could not handle the motivating case anymore. @dtcxzyw, could you test the impact of this patch? If it optimizes other cases still, I wouldn't argue about it.

Reply:
An unrelated idea occurred to me: could we rank `select cond, C1, C2` with the same number as constants in the Reassociate pass?